feat: allow any spl paid token flag

Yolley opened the pull request:
Add `allow_any_spl_paid_token` flag that will bypass check (and takes precedence) for whether token is in `allowed_spl_paid_tokens` when checking whether transaction is paid.

For our use case we need to allow to use any SPL token for payment - the price of the token will be checked by the oracle. An alternative would be to allow `allowed_spl_paid_tokens` be empty, but I think that having this flag set explicitly makes more sense for backwards compatibility and security reasons.

Yolley commented:
@dev-jodee @amilz please take a look, I can't assign reviewers.

dev-jodee commented on crates/lib/src/validator/config_validator.rs:
```diff
@@ -669,6 +688,36 @@ mod tests {
+
+        let _ = update_config(config);
+
+        let rpc_client = RpcClient::new("http://localhost:8899".to_string());
```
Can we use mocked RPC client here for unit testing we don't want to require the local validator to be running.

Yolley replied:
I see, thank you, updated the branch, please check.

dev-jodee commented on crates/lib/src/token/token.rs:
```diff
@@ -231,7 +231,9 @@ impl TokenUtil {
                 };
 
-                if !config.validation.allowed_spl_paid_tokens.contains(&mint_address.to_string()) {
+                if !config.validation.allow_any_spl_paid_token
```
nit: have a function called is_spl_paid_token_valid(..mint..) { the 2 conditions } and use that func instead

amilz reviewed (commented):
Thanks for the PR @Yolley -- a couple of thoughts. 

## 1. Unified SplTokenPaymentConfig

Rather than two potentially conflicting fields in the config, could we do a single one that accepts an allowlist or ALL? I think it'd be a cleaner devex and reduce risk of conflicting configs.

Something like this:

```
  // In config.rs
  pub enum SplTokenPaymentConfig {
      All,                    // Accept any SPL token
      Allowlist(Vec<String>), // Accept only specific tokens
  }

  impl Default for SplTokenPaymentConfig {
      fn default() -> Self {
          SplTokenPaymentConfig::Allowlist(vec![])
      }
  }

  // In ValidationConfig
  pub struct ValidationConfig {
      // ... other fields ...
      pub allowed_spl_paid_tokens: SplTokenPaymentConfig,
  }
```

I think this change might require a tweak to `crates/lib/src/validator/config_validator.rs` to skip check if `All`. Could even add a warning if all is enabled: "Warning allowed for payments" 

## 2. Test coverage--would you mind adding a test or todo to make sure we throw if our oracle does not find a token.

Yolley commented:
> > @amilz I switched to the proposed solution - whilte I am not a fan of it, it's not necessarily bad. Also added a test for jup oracle for when it does not return price for a token. Please take look.
> 
> Hey @Yolley, sorry we've been crunching out features, the pr itself seems fine, I'm just trying to think of a case where a node operator would be willing to allow any spl tokens, with Solana having many 100s of thousands of tokens created every week or so, this seems extremely dangerous, what use case did you have in mind that would warrant allowing all spl tokens as payment?

Hey, our use-case is to use fee relayer when claiming an Airdrop - so sol/rent fees are covered by the relayer, and Airdrop may be created for any token on our platform.

Another use case I'd say could be for token-swaps? Like on a DEX - in case user does not have enough sol to cover processing fees, so the platform allows to pay with swapped tokens.

I don't see how it's "extremely" dangerous tho - we still check for the token price and make sure that the fees are covered according to pricing model configuration. I guess there could be an issue with prices changing a lot, but I don't see a clear fix for this right - I guess we could use a mix of volume/orderbook state or something, but it also depends on what Oracle can provide.

dev-jodee commented:
> @dev-jodee @amilz pretty please, I am tired of rebasing the branch.

Think the pr makes sense, sorry for the recurring rebase, branch is actively being worked on, finally merged the huge pre-release PR into main, if you rebase against main (might be easier to make a new branch from main and cherry-pick your commit into it), I can do final review and merge

Yolley commented on crates/lib/src/config.rs:
```diff
@@ -58,13 +58,61 @@ impl Default for FeePayerBalanceMetricsConfig {
+}
+
+impl SplTokenPaymentConfig {
+    pub fn has_token(&self, token: &str) -> bool {
```
Why tho? This method is specific to this config, config may not even be knowledgeable about whether it for allowed payment tokens or not. This config just stores token configuration, and either has specific tokens, or "all"

Yolley replied:
You're right, renamed it and pushed

Yolley commented:
Hey, @amilz I've added a test `test_initialize_atas_when_all_tokens_are_allowed`, also regarding warning - even if user passes some tokens, node can still experience volatility if they pass volatile tokens, does it really make sense to issue another warning here, given that the configuration to enable all tokens is explicit?
